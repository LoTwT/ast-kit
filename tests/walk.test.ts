import { describe, expect, expectTypeOf, test, vi } from 'vitest'
import {
  type ImportBinding,
  babelParse,
  walkASTSetup,
  walkImportDeclaration,
} from '../src'
import type * as t from '@babel/types'

describe('walk', () => {
  test('walkImportDeclaration', () => {
    const ast = babelParse(
      `
      import { a } from 'a' 
      import * as b from 'b'
      import c from 'c'
      import 'd'
      import e, { f } from 'e'
      import type { g } from 'g'
      import type h from 'h'
      import type * as i from 'i'
      import type j from 'j'
      import { type h } from 'h'`,
      'ts',
    )
    const imports: Record<string, ImportBinding> = {}
    for (const stmt of ast.body) {
      if (stmt.type === 'ImportDeclaration')
        walkImportDeclaration(imports, stmt)
    }

    expect(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      Object.values(imports).map(({ specifier, ...i }) => i),
    ).toMatchInlineSnapshot(`
      [
        {
          "imported": "a",
          "isType": false,
          "local": "a",
          "source": "a",
        },
        {
          "imported": "*",
          "isType": false,
          "local": "b",
          "source": "b",
        },
        {
          "imported": "default",
          "isType": false,
          "local": "c",
          "source": "c",
        },
        {
          "imported": "default",
          "isType": false,
          "local": "e",
          "source": "e",
        },
        {
          "imported": "f",
          "isType": false,
          "local": "f",
          "source": "e",
        },
        {
          "imported": "h",
          "isType": true,
          "local": "h",
          "source": "h",
        },
      ]
    `)
  })

  test('walkASTSetup', async () => {
    const ast = babelParse(
      `
      function foo() {}
      function bar(id: string) {}
      const baz = 1
      `,
      'ts',
    )

    const walkFunctionDeclaration = vi.fn()
    const walkTS = vi.fn()
    const walkIdentifier = vi.fn()

    const p = walkASTSetup(ast, (setup) => {
      setup.onEnter('FunctionDeclaration', (node, parent) => {
        expectTypeOf<t.FunctionDeclaration>(node)
        expectTypeOf<t.ParentMaps['FunctionDeclaration']>(parent)
        walkFunctionDeclaration()
      })
      setup.onEnter(['NumericLiteral', 'TSStringKeyword'], (node, parent) => {
        expectTypeOf<t.NumericLiteral | t.TSStringKeyword>(node)
        expectTypeOf<t.ParentMaps['NumericLiteral' | 'TSStringKeyword']>(parent)
        walkTS()
      })
      setup.onEnter(
        (node): node is t.Identifier => node.type === 'Identifier',
        (node, parent) => {
          expectTypeOf<t.Identifier>(node)
          expectTypeOf<t.ParentMaps['Identifier']>(parent)
          walkIdentifier()
        },
      )

      setup.onEnter((node) => {
        expectTypeOf<t.Node>(node)
        return 123
      })
    })
    expect(walkFunctionDeclaration).toBeCalledTimes(0)

    await p
    expect(walkFunctionDeclaration).toBeCalledTimes(2)
    expect(walkTS).toBeCalledTimes(2)
    expect(walkIdentifier).toBeCalledTimes(4)
  })
})
