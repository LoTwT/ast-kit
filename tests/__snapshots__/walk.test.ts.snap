// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`walk > walkExportDeclaration > ExportAllDeclaration 'type' 1`] = `
[
  {
    "exported": "*",
    "isType": true,
    "local": "*",
    "source": "a",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportAllDeclaration 'value' 1`] = `
[
  {
    "exported": "*",
    "isType": false,
    "local": "*",
    "source": "a",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'class with id' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "Cls",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'class' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'expression with VariableDeclaration' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "one",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'expression' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'function with id' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "foo",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'function' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'ts declaration' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "tsFoo",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '1' 1`] = `
[
  {
    "exported": "v",
    "isType": false,
    "local": "v",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '2' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "f",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '3' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '4' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '5' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "x",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '6' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "v",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '7' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "x",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '8' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "v",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '9' 1`] = `
[
  {
    "exported": "*",
    "isType": false,
    "local": "*",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '10' 1`] = `
[
  {
    "exported": "ns",
    "isType": false,
    "local": "*",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportNamedDeclaration declarations 1`] = `
[
  {
    "exported": "a",
    "isType": false,
    "local": "a",
    "source": null,
  },
  {
    "exported": "b",
    "isType": false,
    "local": "b",
    "source": null,
  },
  {
    "exported": "one",
    "isType": false,
    "local": "one",
    "source": null,
  },
  {
    "exported": "foo",
    "isType": false,
    "local": "foo",
    "source": null,
  },
  {
    "exported": "clz",
    "isType": false,
    "local": "clz",
    "source": null,
  },
  {
    "exported": "enumA",
    "isType": false,
    "local": "enumA",
    "source": null,
  },
  {
    "exported": "interfaceA",
    "isType": true,
    "local": "interfaceA",
    "source": null,
  },
  {
    "exported": "typeA",
    "isType": true,
    "local": "typeA",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportNamedDeclaration specifiers 1`] = `
[
  {
    "exported": "one",
    "isType": false,
    "local": "one",
    "source": null,
  },
  {
    "exported": "aliasOne",
    "isType": false,
    "local": "one",
    "source": null,
  },
  {
    "exported": "specifierTypeOne",
    "isType": true,
    "local": "one",
    "source": null,
  },
  {
    "exported": "d",
    "isType": false,
    "local": "default",
    "source": "z",
  },
  {
    "exported": "nodeTypeOne",
    "isType": true,
    "local": "one",
    "source": null,
  },
  {
    "exported": "two",
    "isType": true,
    "local": "two",
    "source": null,
  },
  {
    "exported": "a",
    "isType": false,
    "local": "a",
    "source": "z",
  },
  {
    "exported": "aliasB",
    "isType": false,
    "local": "b",
    "source": "z",
  },
  {
    "exported": "zAll",
    "isType": false,
    "local": "*",
    "source": "z",
  },
  {
    "exported": "c",
    "isType": true,
    "local": "c",
    "source": "z",
  },
  {
    "exported": "aliasTypeD",
    "isType": true,
    "local": "d",
    "source": "z",
  },
  {
    "exported": "zTypeAll",
    "isType": true,
    "local": "*",
    "source": "z",
  },
]
`;
